make the goal selectable

4383 opened the pull request:
Change-Id: I592fc6d7048d8e927f7a262683dc19163d136eca

4383 committed:
make the goal selectable

Change-Id: I592fc6d7048d8e927f7a262683dc19163d136eca

4383 commented on goals/selected/remove-eventlet.rst:
```diff
@@ -0,0 +1,104 @@
+.. [1] https://issues.redhat.com/browse/OSPRH-5960
+.. [2] https://issues.redhat.com/browse/OSPRH-5980
+.. [3] https://eventlet.readthedocs.io/en/latest/asyncio/migration.html
+.. [4] 
```
All these links depends on patches that are not yet merged and so the documentation page does not yet exists. I'll append them when the eventlet PRs will be merged

4383 commented:
Do not worry too much about all the sections I removed from the "proposed" version of this goal, I'll recycle all of them into the eventlet documentation or in various articles.

I based the new version (the select-able) on the oslo.privsep migration goal, which looks closer to the things that we are trying to do now that we created awaitlet and that we made eventlet compatible with Asyncio. 
https://governance.openstack.org/tc/goals/selected/migrate-to-privsep.html

4383 commented:
@gouthampacha please take a look to this draft when you will have some spare time.
Thanks for your time

4383 commented:
@gouthampacha mind you having a look at this draft?

gouthampacha commented on goals/selected/remove-eventlet.rst:
```diff
@@ -0,0 +1,104 @@
+
+OpenStack is built on the top of asynchronous mechanisms.
+Many Openstack deliverables use the Eventlet library to obtain asynchronous
+those asynchronous mechanismes. Problem, Eventlet lake of active maintainers
+and that led that library to a point where Eventlet is broken at each new
+CPython versions.
```
```suggestion
those asynchronous mechanismes. The problem is that Eventlet lacks active maintainers
and that has led the library to a point where Eventlet is broken with each new
CPython version.
```

gouthampacha commented on goals/selected/remove-eventlet.rst:
```diff
@@ -0,0 +1,104 @@
+and that led that library to a point where Eventlet is broken at each new
+CPython versions.
+
+Asyncio based solutions (awaitlet, aiohttp, etc) offers a flexible, perennial,
```
```suggestion
Asyncio based solutions (awaitlet, aiohttp, etc) offer flexible, perennial,
```

gouthampacha commented on goals/selected/remove-eventlet.rst:
```diff
@@ -0,0 +1,104 @@
+Asyncio based solutions (awaitlet, aiohttp, etc) offers a flexible, perennial,
+and modern alteratives to Eventlet. Since the Eventlet Asyncio hub was
+released, allowing Eventlet and Asyncio to be ran into the same process, the
+community has encouraged the usage of those alternatives and the deprecation
```
suggest: Eventlet community

gouthampacha commented on goals/selected/remove-eventlet.rst:
```diff
@@ -0,0 +1,104 @@
+
+  eventlet-removal
+
+A new sub-task in middle term [1]_ or long term [2]_ stories will be created
```
s/middle/short

or maybe "medium"

gouthampacha commented on goals/selected/remove-eventlet.rst:
```diff
@@ -0,0 +1,104 @@
+===================
+
+#. Get the oslo deliverables doing networkIO providing Asyncio based drivers
+   and backends (oslo.cache, oslo.messaging, oslo.db, etc);
```
is it possible to enumerate "etc" here? 

IIRC you had already compiled a list..

gouthampacha commented on goals/selected/remove-eventlet.rst:
```diff
@@ -0,0 +1,104 @@
+#. Get the oslo deliverables doing networkIO providing Asyncio based drivers
+   and backends (oslo.cache, oslo.messaging, oslo.db, etc);
+#. Get libraries like OpenStackSDK migrated;
+#. Get a reference user project elected;
```
What does this mean?

4383 replied:
That was suggested by Pete, to check the Oslo new "async" drivers/backends in a chosen Openstack deliverable:
https://review.opendev.org/c/openstack/governance/+/902585/comment/80568a67_d1db7f48/

I can update that sentence to better explain that point.

gouthampacha commented on goals/selected/remove-eventlet.rst:
```diff
@@ -0,0 +1,104 @@
+   and backends (oslo.cache, oslo.messaging, oslo.db, etc);
+#. Get libraries like OpenStackSDK migrated;
+#. Get a reference user project elected;
+#. Get non actively maintained deliverables retired;
```
s/deliverables/OpenStack deliverables

gouthampacha commented on goals/selected/remove-eventlet.rst:
```diff
@@ -0,0 +1,104 @@
+#. Get all other OpenStack deliverables relying on Eventlet migrated;
+#. Get usage of Eventlet in oslo deliverables removed;
+#. Get Eventlet retired from OpenStack;
+#. Get Eventlet abandoned or bequeath to someone else.
```
the last point wouldn't be necessary in this goal..

gouthampacha commented on goals/selected/remove-eventlet.rst:
```diff
@@ -0,0 +1,104 @@
+
+The main reference guide could be found in [3]_. This guide provides
+information about how to migrate from Eventlet to the available alternatives.
+As commented in the first section, if the alternative selected to migrate
```
s/commented/noted

gouthampacha reviewed (commented):
Hi @4383 - super late to this review; sorry about this.. i've some questions/minor comments inline.. 

Some things that would help this:

1) Timeline
2) How/Who can help with this effort

Regarding (1); i think you can break down the completion criteria into phases and suggest OpenStack releases that you aim for these to be completed within.. This isn't set in stone - and can always change.. but, it provides reviewers with a chance to picture how long this will take. 

Regarding (2), I spoke to a few people about pop up teams and gathered the know how. I can propose a change to create the group. We can have discussions across the community on an etherpad and have different folks share their progress with the goal, or collaborate to resolve questions.. does this sound fine to you?

4383 commented:
> Regarding (2), I spoke to a few people about pop up teams and gathered the know how. I can propose a change to create the group. We can have discussions across the community on an etherpad and have different folks share their progress with the goal, or collaborate to resolve questions.. does this sound fine to you?

Excellent!
I think we also have to consider the downstream side of this topic, so, I recently had a discussion with some project managers, and they advised me to join the project managers meetings on a regular manner to "marketize" this topic and get help about prioritizing this topic on their teams. Hence, PMs will spread the words and manage this removal into their own teams by allocating resources etc...

In parallel of that, I started to create a tool that migrate an existing Eventlet code into one of proposed alternatives (alternatives proposed in the original goal proposal). This tool will be able to fully migrate a full deliverable (openstack deliverable or not) in one shot, or to only migrate a specific module, or, again, to assist its user step by step and file by file. This will lower the cost of the migration. Rather than simply providing a migration guide and ask people to manually migrate their deliverables by following this guide, I think we can provide this tool which is based on the algorithm defined in the guide. I already started to create that tool, I made good progress, I think I'll be able to show some demo in the coming weeks.

4383 commented:
I updated this goal version with your previous comments.
Please, feel free to review it again.

gouthampacha commented on goals/selected/remove-eventlet.rst:
```diff
@@ -0,0 +1,109 @@
+
+OpenStack is built on the top of asynchronous mechanisms.
+Many Openstack deliverables use the Eventlet library to obtain asynchronous
+those asynchronous mechanismes. The problem is that Eventlet lacks of active
```
```suggestion
those asynchronous mechanisms. The problem is that Eventlet lacks of active
```

typo

gouthampacha commented on goals/selected/remove-eventlet.rst:
```diff
@@ -0,0 +1,109 @@
+Many Openstack deliverables use the Eventlet library to obtain asynchronous
+those asynchronous mechanismes. The problem is that Eventlet lacks of active
+maintainers and that has led the library to a point where Eventlet is broken
+with each new CPython versions.
```
```suggestion
with each new CPython version.
```
typo

gouthampacha commented on goals/selected/remove-eventlet.rst:
```diff
@@ -0,0 +1,109 @@
+
+Asyncio based solutions (awaitlet, aiohttp, etc) offer flexible, perennial,
+and modern alteratives to Eventlet. Since the Eventlet Asyncio hub was
+released, allowing Eventlet and Asyncio to be ran into the same process, the
```
```suggestion
released, allowing Eventlet and Asyncio to be run within the same process, the
```

grammar

gouthampacha commented on goals/selected/remove-eventlet.rst:
```diff
@@ -0,0 +1,109 @@
+Completion Criteria
+===================
+
+#. (2025-1 or 2025-2) Get the oslo deliverables doing networkIO providing
```
```suggestion
#. (2025.1 or 2025.2) Get the oslo deliverables doing networkIO providing
```

gouthampacha commented on goals/selected/remove-eventlet.rst:
```diff
@@ -0,0 +1,109 @@
+#. (2025-2) Get non actively maintained Openstack deliverables retired;
+#. (2026-2) Get all other OpenStack deliverables relying on Eventlet migrated;
+#. (2027-1) Get usage of Eventlet in oslo deliverables removed;
+#. (2027-2) Get Eventlet retired from OpenStack;
```
```suggestion
#. (2025.1) Get libraries like OpenStackSDK migrated;
#. (2025.2) Get an OpenStack deliverable as reference user project selected to
   test Oslo changes related to new Asyncio based drivers;
#. (2025.2) Get non actively maintained OpenStack deliverables retired;
#. (2026.2) Get all other OpenStack deliverables relying on Eventlet migrated;
#. (2027.1) Get usage of Eventlet in oslo deliverables removed;
#. (2027.2) Get Eventlet retired from OpenStack;
```

gouthampacha commented on goals/selected/remove-eventlet.rst:
```diff
@@ -0,0 +1,109 @@
+substitution.
+
+As commented in the OpenStack mailing list [9]_, the migration is based on the
+freedom of decision-making. Some Openstack deliverables represent thousand of
+line code. Asyncio is by nature invasive. Using Asyncio everywhere would
```
```suggestion
freedom of decision-making. Some Openstack deliverables represent thousands of
lines of code. Asyncio is by nature invasive. Using Asyncio everywhere would
```

gouthampacha commented on goals/selected/remove-eventlet.rst:
```diff
@@ -0,0 +1,109 @@
+As commented in the OpenStack mailing list [9]_, the migration is based on the
+freedom of decision-making. Some Openstack deliverables represent thousand of
+line code. Asyncio is by nature invasive. Using Asyncio everywhere would
+require to rewrite entirely the deliverables. For this reason we created
```
```suggestion
require to rewrite the deliverables entirely. For this reason we created
```

gouthampacha commented on goals/selected/remove-eventlet.rst:
```diff
@@ -0,0 +1,109 @@
+==================================
+
+Progress is maintained on the public tracker links below:
+* short term actions (done): [11]_
+* middle term actions: [1]_
+* long term actions: [2]_
```
Unfortunately, I feel the community wouldn't take to using Red Hat Jira as a tracker for this work; wiki.openstack.org or etherpad.opendev.org may be good alternatives if this doesn't fit traditional tools like Launchpad or OpenDev Storyboard. 

This can be something you can delegate the Pop Up team to track though.. but, early heads up that when we post this patch to gerrit, this will invite some discussion

4383 replied:
No problem about your early warning, you are right to clarify this.
My problem is that I've to track all the things related to this migration into jira, so doing another tracking over the upstream tools duplicate the workload and the bureaucracy stuff.

If you think is it possible to delegate that synchronization tracking to the pop up team, then I won't say no :)

I might refactor that document to only speak about the upstream tools (wiki or etherpad) and perhaps we could discuss the jira sync only into the internal discussions of the pop up team, thoughts?
Going this way would reduce the noise around the discussion for goal status update (proposed to selected).

gouthampacha reviewed (commented):
Thank you for clarifying the timeline! This looks great; some comments inline

4383 commented:
Thanks for your other comments, I'll address them soon.

4383 commented:
@gouthampacha: I just pushed a new patch set where I addressed all your comments. Please have a look.

I also decided to remove some links mostly related to pull requests I made on Eventlet and where I seen no progress on reviews from my co-maintainers. Lets remains on the facts only.

In parallel of that, I also made really good progress with my tool to automatically migrate Eventlet usages. Indeed I wrote several algorithms and logics around that:

1. I wrote a tool that create a snapshot of the public API of any existing python library, so we are able to get a snapshot for Eventlet and for Asyncio, AIOHtttp, Awaitlet etc...
2. I wrote a second tool that take in entry 2 or more snapshots and that apply a classification algorithms like DBSCAN and that generate a migration schema at the config format (yaml, toml,etc..)
3. I designed a third tool that take for entry this migration schema and a local path toward an application to migrate. This tool modify the AST of the this application and apply the changes to all the files in the specified directory.
4. and finally I desgned a last tool that check that the code is executable and that async functions functions of the code are ok.

With that we will be able to create scenario for any migration we want between 2 or more libraries. It would be possible to say : I want to replace eventlet by asyncio and I want to give the priority to aiohttp for all the the usage of wsgi module of eventlet. 

In this previous example AIOHttp will be used in priority over Asyncio, and Asyncio will be used for all the remaining use cases of eventlet usages.

This toolchain should really lower the cost of our migration.
As this toolchain is designed to create plugglable migration scenarios, this toolchain will be also available for all the future use case similar to the eventlet removal. By example the migration of oslo.service toward a facade for cotyledon and futurist.

I'll share you the links when I move this toolchain to public.
